src/controller/ICDRegistrationDelegate.h

```diff
@@ -0,0 +1,62 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *    This file contains definitions for DeviceProxy for a device that's undergoing
+ *    commissioning process. The objects of this will be used by Controller applications
+ *    to interact with the device. The class provides mechanism to construct, send and receive
+ *    messages to and from the corresponding CHIP devices.
+ */
+
+#pragma once
+
+#include <cinttypes>
+
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace Controller {
+
+class ICDRegistrationDelegate
+{
+public:
+    static constexpr size_t KeySize = 16;
+    using ICDKey                    = FixedByteSpan<KeySize>;
+
+    typedef void (*OnSymmetricKeyGenerationCompleted)(void * context, CHIP_ERROR error, NodeId controllerNodeId, uint64_t subjectId,
+                                                      ICDKey key, Optional<ICDKey> verificationKey);
+
+    virtual ~ICDRegistrationDelegate() = default;
+
+    /**
+     * @brief
+     *   Called when the operational commissioning completed, the controller may register it with the ICD device.
+     * The delegate can generate the symmetric key and call commissioner->OnSymmetricKeyGenerated to complete ICD
+     * registration.
+     *   This delegate supports Commissioner Self-Registration and External Controller Registration flow.
+     * @param[in] commissioner The commissioner used to commission the device.
+     * @param[in] deviceId     The peer id of the device.
+     */
+    virtual void GenerateSymmetricKey(NodeId device, Callback::Callback<OnSymmetricKeyGenerationCompleted> * onCompletion) = 0;
+
+    virtual void ReadyForSubscription(NodeId device, uint32_t icdCounter) = 0;
```

Comment:
when are you going to invoke ReadyForSubscription callback? just put the note: subscription only happens over case, so if you start the subscription after you receive the registerDevice response over pase, subscription coud not work..